src/main/scala/edu/cmu/cs/obsidian/parser/AST.scala

```diff
@@ -48,79 +58,266 @@ sealed abstract class InvokableDeclaration() extends Declaration {
 /* Expressions */
 case class ReferenceIdentifier(name: String) extends Expression {
     override val toString = name
+
+    override def substitute(genericParams: Seq[GenericType], actualParams: Seq[ObsidianType]): ReferenceIdentifier = this
 }
 
-case class NumLiteral(value: Int) extends Expression
-case class StringLiteral(value: String) extends Expression
-case class TrueLiteral() extends Expression
-case class FalseLiteral() extends Expression
+case class NumLiteral(value: Int) extends Expression {
```

Comment:
This is a lot of duplicated code. The following might be a little better…

````
abstract case class BinaryExpression(e1: Expression, e2: Expression) extends Expression {
    def make(newE1: Expression, newE2: Expression): BinaryExpression

    override def substitute(genericParams: Seq[GenericType], actualParams: Seq[ObsidianType]): BinaryExpression = {
        make(e1.substitute(genericParams, actualParams), e2.substitute(genericParams, actualParams)).setLoc(this)
    }
}

case class Conjunction(override e1: Expression, override e2: Expression) extends BinaryExpression(e1, e2) {
    def make(newE1: Expression, newE2: Expression): BinaryExpression = Conjunction(newE1, newE2)
}
````
etc.

Reply:
I wonder if there's a Scala mechanism that would let us cut out all those `make` methods, but if so, it's not apparent to me what it is.